### PR TITLE
CSP-472 - MembersRegion table and Validation middleware

### PR DESCRIPTION
**Pull request by LiamWright**
Added MembersRegion table/entities

### CODE REVIEW

File: src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿CREATE TABLE [dbo].[MemberRegion]
+(
+	[MemberId] UNIQUEIDENTIFIER NOT NULL PRIMARY KEY,
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql:3]: Primary key should be composite of MemberId and RegionId

File: src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs [hunk 2]
```diff
@@ -19,6 +19,25 @@ public CreateMemberCommandValidator()
                 .Matches(Constants.RegularExpressions.EmailRegex);
             RuleFor(x => x.Joined)
                 .NotEmpty();
+            RuleFor(x => x.Regions).Custom((regions, context) =>
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs:22]: This should be an async validator checking against the database records.

File: src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs [hunk 3]
```diff
@@ -1,9 +1,9 @@
 ﻿using FluentValidation;
-using SFA.DAS.AANHub.Application.Common.Commands;
+using SFA.DAS.AANHub.Application.Commands.CreateMember;
 
 namespace SFA.DAS.AANHub.Application.Common.Validators
 {
-    public class CreateMemberCommandValidator : AbstractValidator<BaseMemberCommand>
+    public class CreateMemberCommandValidator : AbstractValidator<CreateMemberCommand>
     {
         public CreateMemberCommandValidator()
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs:8]: We need a reusable validator that can be included in all the Create commands. To do this, the validator should be on an interface, which is then implemented by each create command. See validator example [here](https://github.com/SkillsFundingAgency/das-roatp-api/blob/fe881635ace7209e614fbca6145be539be0793b1/src/SFA.DAS.Roatp.Application/Common/UkprnValidator.cs#L6)  and the usage example [here](https://github.com/SkillsFundingAgency/das-roatp-api/blob/fe881635ace7209e614fbca6145be539be0793b1/src/SFA.DAS.Roatp.Application/ProviderCourse/Commands/CreateProviderCourse/CreateProviderCourseCommandValidator.cs#L28)

**Review by DevenKShah — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql (hunk 1 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql]: Done

File: src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs (hunk 2 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs]: Implemented

File: src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs (hunk 3 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Common/Validators/CreateMemberCommandValidator.cs]: I have abstrated the base validation into `IBaseMemberCommand/BaseMemberCommand`, and introduced this to `CreateMemberCommandValidator`

File: src/SFA.DAS.AANHub.Database/Tables/Audit.sql [hunk 4]
```diff
@@ -2,7 +2,7 @@
 (
     [Id] BIGINT NOT NULL IDENTITY(1, 1) PRIMARY KEY, 
     [AuditTime] DATETIME2 NOT NULL, 
-    [ActionedBy] UNIQUEIDENTIFIER NOT NULL, 
+    [ActionedBy] NVARCHAR(256) NOT NULL,
```
Comment (LiamWright) [inline src/SFA.DAS.AANHub.Database/Tables/Audit.sql:5]: Changed as errors were thrown when auditing with same memberId

Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Database/Tables/Audit.sql:5]: There should not be any constraint errors as we have no foreign key to member, so what was it?

Comment (LiamWright) [reply src/SFA.DAS.AANHub.Database/Tables/Audit.sql]: Apologies, this is a misnomer. The error that I saw was unrelated to this. `uniqueidenifier` has been re-instated.

File: src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿CREATE TABLE [dbo].[MemberRegion]
+(
+	[MemberId] UNIQUEIDENTIFIER NOT NULL, 
+    [RegionId] INT NOT NULL
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql:4]: Alignment

File: src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+	PRIMARY KEY (MemberId, RegionId)
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql:5]: Add primary key name `PK_MemberRegion`

Comment (LiamWright) [reply src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql:5]: image

The formatting seems to just be a Github issue than a formatting issue

File: src/SFA.DAS.AANHub.Application.UnitTests/Common/Validators/CreateMemberCommandValidatorTests.cs [hunk 7]
```diff
@@ -1,20 +1,38 @@
...
     {
+        private readonly Mock<IRegionsReadRepository> _regionsReadRepository;
+
+        private readonly List<Region> _regions = new()
+        {
+            new Region { Id = 1, Area = "Area1", Ordering = 1 },
+            new Region { Id = 2, Area = "Area2", Ordering = 2 }
+        };
+
+        public CreateMemberCommandValidatorTests()
```
[the remark this answers is not in the document]
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application.UnitTests/Common/Validators/CreateMemberCommandValidatorTests.cs]: Moved to region tests

File: src/SFA.DAS.AANHub.Application/Common/Validators/BaseMemberValidator.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            RuleFor(x => x.Joined)
+                .NotEmpty();
+            RuleFor(x => x.Regions).MustAsync(async (regions, cancellationToken) =>
+            {
+                if (regions == null) return true;
+
+                var dbRegions = await regionsReadRepository.GetAllRegions();
+                var regionList = dbRegions.Select(item => item.Id).ToList();
+
+                return regions.Any(region => regionList.Contains(region));
```
[the remark this answers is not in the document]
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Common/Validators/BaseMemberValidator.cs]: Updated to `.All()`, and tests adjusted

File: src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql (hunk 5 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Database/Tables/MemberRegion.sql:4]: image

The formatting seems to just be a Github issue than a formatting issue

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**